File: C++/Medium/WrongAnswer/sell-diminishing-valued-colored-balls_6.cpp
class Solution {
public:
    int maxProfit(vector<int>& inventory, int orders) {
        
        int mod = 1000000007;
        long long sum = 0;
        sort(inventory.begin(), inventory.end(), greater<int>());
        int i = 0;
        while (orders > 0) {
            int cur = inventory[i];
            int next = i < inventory.size() - 1 ? inventory[i + 1] : 0;
            int cnt = next == cur ? 0 : orders / (i + 1);
            if (cnt > 0) {
                int step = cur - next;
                if (cnt > step) {
                    sum += (long long)(cur + next + 1) * step / 2 * (i + 1);
                    orders -= step * (i + 1);
                    cur = next;
                } else {
                    sum += (long long)(cur + cur - cnt + 1) * cnt / 2 * (i + 1);
                    orders -= cnt * (i + 1);
                    cur -= cnt;
                }
            } else {
                sum += (long long)cur * orders;
                orders = 0;
            }
            sum %= mod;
            i++;
        }
        return sum;
    }
};

File: Java/Medium/Accepted/sell-diminishing-valued-colored-balls_6.java
class Solution {
    public int maxProfit(int[] inventory, int orders) {
        
        long mod = 1000000007;
        Arrays.sort(inventory);
        int n = inventory.length;
        long ans = 0;
        int i = n - 1;
        while (orders > 0) {
            while (i > 0 && inventory[i] == inventory[i - 1]) i--;
            long cnt = n - i;
            long diff = inventory[i] - (i > 0 ? inventory[i - 1] : 0);
            if (diff * cnt <= orders) {
                ans += (inventory[i] + inventory[i] - diff + 1) * diff * cnt / 2;
                orders -= diff * cnt;
            } else {
                long q = orders / cnt;
                long r = orders % cnt;
                ans += (inventory[i] + inventory[i] - q + 1) * q * cnt / 2;
                ans += r * (inventory[i] - q);
                orders = 0;
            }
            i--;
            ans %= mod;
        }
        return (int)ans;
    }
}